proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,820 @@
+supervisor (this can be detected via the OIDC discovery document).  To limit the chance that `cluster-001` results in
+audience confusion during the transition period, the token exchange API would not include it for the `pinniped-cli` when
+a dynamic OAuth client is defined with that name (or even more strictly, if any dynamic OAuth client is defined - i.e.
+using the new dynamic OAuth client feature would force `JWTAuthenticator` config updates - we can detect if any dynamic
+OAuth client is defined via a live list call with limit=1).
+
+###### Option 2: Require a static, reserved prefix for all dynamic OAuth client IDs
+
+When a dynamic client is configured, the full name will not be configurable.  Instead, only the suffix will be
+configurable and the suffix will not be allowed to contain `--`.  The prefix will be always be
+`client.oauth.pinniped.dev--`.  Thus if the suffix is set to `fancy-client`, the ID for the client will be
+`client.oauth.pinniped.dev--fancy-client`.  The token exchange API will be updated to reject attempts to request an
+audience with this prefix. It may be prudent to rename the CLI OAuth client to `client.oauth.pinniped.dev--pinniped-cli`
+and prevent a token exchange for the `pinniped-cli` audience.  `JWTAuthenticator` will be updated to reject audiences
+with the `client.oauth.pinniped.dev--` prefix as well as `pinniped-cli` to disallow use of ID tokens issued during the
+authorization flow.  Users will need to understand this prefixing rule as they will need to configure their app to use
+`client.oauth.pinniped.dev--fancy-client` as the client ID.  Renaming the `pinniped-cli` would be phased effort with the
+old name being supported for some number of releases.  The supervisor would issue warnings to indicate that kubeconfigs
+need to be regenerated to use the new client name (dynamic clients would not be allowed to use this reserved name).
+
+###### Option 2.5: Some mix of Option 1 and 2
+
+It is likely possible to use both options in conjunction.  Note that mixing old and new versions of the supervisor and
+concierge may still result in audience confusion as components may not enforce all checks.
+
+###### Option 3: Break apart signing keys for dynamic OAuth clients
+
+Cryptographic isolation of signing keys is a guaranteed way to prevent any form of confusion between ID tokens issued
+by the authorization code flow and and ID tokens issued by the token exchange API.  At a high level, we could separate
+the single OIDC issuer into as many as four distinct issuers (authorization code vs. token exchange x static vs. dynamic
+OAuth client).  Consumers would be configured to trust the exact sub-issuer that they are expected to understand.  The
+OIDC discovery protocol makes this incredibly painful to express.  This technical specifics for this option have been
+mostly omitted as it does not seem viable.
+
+###### Option 4: Migrate to token webhook
```

Comment:
An important bit that I forgot to include.

Two scenarios that we should aim to prevent:

1. The use of ID tokens from the auth code flow directly against the API server or used via the `JWTAuthenticator` (this is not a real problem currently because we only have the `pinniped-cli` today, but once we allow for dynamic clients, an admin could try to use dynamic clients to represent clusters - like they would with Dex - instead of using the token exchange API, or worse, they could try to re-use the dynamic clients across multiple clusters like they could with Dex)
2. Accidental use of the above by a "legacy" `JWTAuthenticator` (i.e. one configured for the existing pinniped TCR flow - this is partially described above as the audience confusion item `1` but is more referring to existing deployments of pinniped that are using the `JWTAuthenticator` today and, even without upgrading old workload clusters, should be retained as a safe configuration)

An approach to prevent the above while still allowing web apps to learn the identity of the user:

- Nest the `username` and `groups` claims in the auth code flow ID token under a new `pinniped` object and thus make it impossible for the `JWTAuthenticator`/Kube OIDC integration from consuming these ID tokens since that code lacks the ability to read into nested fields (i.e. force them to use the cluster scoped webhook tokens)
- Further gate the inclusion of the new `pinniped` object in the auth code flow ID token behind a new `profile` scope (this would not be requested by the CLI)